File: src/main/java/xyz/kazuthecat/coffeebot/commands/CoffeeLog.java
package xyz.kazuthecat.coffeebot.commands;

import com.jagrosh.jdautilities.command.Command;
import com.jagrosh.jdautilities.command.CommandEvent;
import net.dv8tion.jda.api.entities.User;
import xyz.kazuthecat.coffeebot.DBHandler;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class CoffeeLog extends Command {
    private final DBHandler dbHandler;
    private final List<User> initalizedUsers;

    public CoffeeLog(DBHandler dbHandler) {
        this.name = "coffeelog";
        this.help = "logs a cup of coffee";
        this.guildOnly = false;
        this.dbHandler = dbHandler;
        this.initalizedUsers = new ArrayList<>();

        String coffeelog_tbl = "CREATE TABLE IF NOT EXISTS coffeelog(" +
                "id      BIGINT NOT NULL, " +
                "cups    INT NOT NULL, " +
                "PRIMARY KEY (id) );";

        String coffeetypes_tbl = "CREATE TABLE IF NOT EXISTS coffeetypes(" +
                "id      BIGINT NOT NULL, " +
                "name    VARCHAR(30) NOT NULL, " +
                "cups    INT NOT NULL, " +
                "PRIMARY KEY (id, name) );";

        dbHandler.execute(new String[]{coffeelog_tbl, coffeetypes_tbl});
    }

    @Override
    protected void execute(CommandEvent event) {
        String[] arglist = event.getArgs().split(" ");

        User user = event.getAuthor();
        String uid = user.getId();
        if (!initalizedUsers.contains(user)) {
            dbHandler.execute(
                    "INSERT IGNORE INTO coffeelog (id,cups) VALUES (?, 0);",
                    new String[][]{{uid}}
                    );
            initalizedUsers.add(user);
        }

        String reply;
        switch (arglist[0]) {
            case "add":   reply = addCups(arglist, uid); break;
            case "check": reply = checkCups(uid); break;
            default:      reply = "That's not a valid subcommand."; break;
        }
        event.reply(reply);
    }

    private String addCups(String[] arglist, String uid) {
        String numCups = "1";
        boolean cupOverflow = false;
        if (arglist.length > 1 && arglist[1].matches("\\d+")) {
            numCups = arglist[1];
            if (Integer.parseInt(numCups) > 10) {
                numCups = "10"; // Max number to avoid the cup number overflowing.
                cupOverflow = true;
            }
        }
        boolean result = dbHandler.execute(
                "UPDATE coffeelog SET cups = cups + ? WHERE id = ?;",
                new String[][]{{numCups, uid}}
                );

        if (result && cupOverflow) {
            return String.format("Success! I have added %s to your cup count! This is the maximum number of cups you can add at one time. :coffee:", numCups);
        } else if (result) {
            return String.format("Success! I have added %s to your cup count! :coffee:", numCups);
        } else {
            return "An error occurred, your cup count probably hasn't been updated. But who knows? :shrug:";
        }
    }

    private String checkCups(String uid) {
        String reply;
        List<Map<String, String>> resultSet = dbHandler.query("SELECT * FROM coffeelog WHERE id = ?", new String[]{uid});
        if (resultSet != null && resultSet.size() != 0) {
            String numCups = resultSet.get(0).get("cups");
            reply = String.format("You've had %s cups of coffee to date! :coffee:", numCups);
        } else {
            reply = "Something went wrong when trying to retrieve the number of cups. :shrug:";
        }
        return reply;
    }
}

File: src/main/kotlin/commands/CoffeeLog.kt
package xyz.kazuthecat.coffeebot.commands

import com.jagrosh.jdautilities.command.Command
import com.jagrosh.jdautilities.command.CommandEvent
import net.dv8tion.jda.api.entities.User
import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.transactions.transaction
import xyz.kazuthecat.coffeebot.ExposedDBHandler

class CoffeeLog (dbHandler: ExposedDBHandler) : Command() {
  // coffeelog is used with Exposed to create/handle/query the database
  // when creating the table the name of the class will be used, which is why it's all lower-case
  private val initializedUsers = mutableMapOf<Long, Int>()
  object coffeelog : Table() {
    val id = long("id").primaryKey()
    val cups = integer("cups")
  }

  init {
    name = "coffeelog"
    help = "logs a cup of coffee"
    aliases = arrayOf("cl")
    guildOnly = false

    dbHandler.getDBConnection()
    transaction {
      SchemaUtils.create(coffeelog)
      coffeelog.selectAll().forEach { initializedUsers[it[coffeelog.id]] = it[coffeelog.cups] }
    }
  }

  override fun execute(event: CommandEvent?) {
    if (event == null) { return }

    val args = event.args.split(" ")
    val author = event.author
    val reply : String
    reply = when (args[0]) {
      "add" -> addCups(args, author)
      "check" -> checkCups(author)
      else -> "I'm not sure what you want me to do ${author.asMention}"
    }
    event.reply(reply)
  }

  private fun addCups(args : List<String>, user : User) : String {
    val uid = user.idLong
    val oldCups = initializedUsers[user.idLong]

    var maxCups = false // Did user try to increase by more than 10?
    var newCups = 1     // How many cups does the user want to add?
    if (args.size > 1 && args[1].matches(Regex("\\d+"))) {
      try {
        newCups = args[1].toInt()
        if (newCups > 10) { newCups = 10; maxCups = true }
      } catch (_ : Exception) {}
    }

    transaction {
      if (oldCups == null) {
        // Initialize user if they're not in the database yet.
        coffeelog.insert { it[id] = uid; it[cups] = newCups }
        initializedUsers[uid] = newCups
      } else {
        // Update old entry if such an entry exists
        initializedUsers[uid] = oldCups + newCups
        coffeelog.update({coffeelog.id eq uid}) {
          it[cups] = oldCups + newCups
        }
      }
    }

    return if (!maxCups) {
      "Added $newCups to your cup count, ${user.asMention}. You're now at ${initializedUsers[uid]}"
    } else {
      "You're only allowed to add 10 cups at a time, so I added 10 to your cup count, ${user.asMention}. " +
        "You're now at ${initializedUsers[uid]}"
    }
  }

  private fun checkCups(user : User) : String {
    val currentCups = initializedUsers[user.idLong]
    return if (currentCups == null) {
      "${user.asMention} You've had 0 cups of coffee to date!"
    } else {
      "${user.asMention} You've had $currentCups cups of coffee to date!"
    }
  }
}
